docs/architecture/adr-003-subaccounts.md

```diff
@@ -0,0 +1,88 @@
+# ADR 3: Module SubAccounts
+
+## Changelog
+
+## Context
+
+Currently `ModuleAccount`s must be declared upon supply keeper initialization. In addition to this they don't allow for separation of fungible coins within an account.
+
+New structs should be added so a `ModuleAccount` can dynamically add accounts.
+
+## Decision
+
+Add the following structs into `x/supply`.
+
+### Implementation Changes
+
+Introduce two new structs:
+
+* `ModuleMultiAccount`
+* `SubAccount`
+
+`ModuleMultiAccount` maintains an array of `SubAccount` as well as a list of permissions defined upon initialization of supply keeper (permAddrs).
+It has no pubkey and no limit on the number of `SubAccount`s that can be created.
+Its constructor returns a `ModuleMultiAccount` with no `SubAccount`s.
+`MultiModuleAccount` will implement the Account interface.
+
+```go
+// SetCoins will return an error to prevent ModuleMultiAccount address from having a balance.
+// SubAccounts can only be appended to the SubAccount array.
+// Passively tracks the sum of all SubAccount balances.
+type ModuleMultiAccount struct {
+    Subaccs []SubAccount
+    Permissions []string
+    Coins sdk.Coins // passively track all sub account balances
+
+    CreateSubAccount(pubkey, address) int // returns id of subaccount
+    GetSubAccount(id int) SubAccount
+}
+```
+
+To invalidate a `SubAccount` the `ModuleMultiAccount` calls `SetAccountDisabled` for a `SubAccount`
+
+```go
+// Implements the Account interface. Address is the ModuleMultiAccount address with the id appended.
+// Permissions must be a subset of its ModuleMultiAccount permissions.
+// A disabled account can do withdraws, but cannot recieve any coins.
+type SubAccount struct {
+    Address sdk.AccAddress // MultiAccount (parent) address with index appended
+    ID uint // index of subaccount
+    Permissions []string
+    Disabled bool
+
+    SetAccountDisabled()
```

Comment:
Is there an advantage to keeping an account disabled opposed to removing it entirely?

Reply:
I believe the idea is that disabled accounts can have their funds passively withdrawn from but will not accept new funds. 
@colin-axner could you confirm? Could you also give an example of how this might be useful?

Reply:
@AdityaSripal that is correct. In relation to coinswap, suppose governance decides to deprecate a trading pair (maybe the coin liquidity is being merged with another pegged equivalent asset), there would need to be some way to enforce that no more deposits could occur. 

This could be abstracted to the handler/keeper level if we want to keep the SubAccount as simple as possible

Reply:
I'm thinking abstracting this to be a application developer decision is the way to go. I will remove the disabling, and add a function for removing accounts, but specify that the account must be empty before it can be removed

Reply:
I decided to leave out the ability to remove accounts. Trying to remove an account from the `ModuleAccount` array would cause indexing not to match up with account numbers. I think this is something we can leave developers to indirectly remove accounts by not referencing them if needed. If removing sub-accounts is a desired feature, I think and issue should be opened and a follow up pr should address the problem

Reply:
Makes sense. Thanks!